lib/modules/manager/terraform/lockfile/index.spec.ts

```diff
@@ -910,4 +910,183 @@ describe('modules/manager/terraform/lockfile/index', () => {
     });
     expect(result).toBeNull();
   });
+
+  it('preserves constraints when current value and new value are same', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(codeBlock`
+      provider "registry.terraform.io/hashicorp/aws" {
+        version     = "3.0.0"
+        constraints = "~> 3.0.0"
+        hashes = [
+          "aaa",
+          "bbb",
+          "ccc",
+        ]
+      }
+    `);
+    fs.findLocalSiblingOrParent.mockResolvedValueOnce('.terraform.lock.hcl');
+
+    mockHash.mockResolvedValueOnce([
+      'h1:lDsKRxDRXPEzA4AxkK4t+lJd3IQIP2UoaplJGjQSp2s=',
+      'h1:6zB2hX7YIOW26OrKsLJn0uLMnjqbPNxcz9RhlWEuuSY=',
+    ]);
+
+    const result = await updateArtifacts({
+      packageFileName: 'main.tf',
+      updatedDeps: [
+        {
+          depName: 'aws',
+          depType: 'provider',
+          packageName: 'hashicorp/aws',
+          registryUrls: ['https://registry.example.com'],
+          newVersion: '3.36.1',
+          currentValue: '~> 3.36',
+          newValue: '~> 3.36',
+        },
+      ],
+      newPackageFileContent: '',
+      config,
+    });
+
+    expect(result).toEqual([
+      {
+        file: {
+          contents: codeBlock`
+            provider "registry.terraform.io/hashicorp/aws" {
+              version     = "3.36.1"
+              constraints = "~> 3.0.0"
+              hashes = [
+                "h1:lDsKRxDRXPEzA4AxkK4t+lJd3IQIP2UoaplJGjQSp2s=",
+                "h1:6zB2hX7YIOW26OrKsLJn0uLMnjqbPNxcz9RhlWEuuSY=",
+              ]
+            }
+          `,
+          path: '.terraform.lock.hcl',
+          type: 'addition',
+        },
+      },
+    ]);
+
+    expect(mockHash.mock.calls).toEqual([
+      ['https://registry.example.com', 'hashicorp/aws', '3.36.1'],
+    ]);
+  });
+
+  it('replaces current value to new version within a constraint', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(codeBlock`
+      provider "registry.terraform.io/hashicorp/aws" {
+        version     = "3.0.0"
+        constraints = "~> 3.0.0"
+        hashes = [
+          "aaa",
+          "bbb",
+          "ccc",
+        ]
+      }
+    `);
+    fs.findLocalSiblingOrParent.mockResolvedValueOnce('.terraform.lock.hcl');
+
+    mockHash.mockResolvedValueOnce([
+      'h1:lDsKRxDRXPEzA4AxkK4t+lJd3IQIP2UoaplJGjQSp2s=',
+      'h1:6zB2hX7YIOW26OrKsLJn0uLMnjqbPNxcz9RhlWEuuSY=',
+    ]);
+
+    const result = await updateArtifacts({
+      packageFileName: 'main.tf',
+      updatedDeps: [
+        {
+          depName: 'aws',
+          depType: 'provider',
+          packageName: 'hashicorp/aws',
+          registryUrls: ['https://registry.example.com'],
+          newVersion: '3.37.0',
+          currentValue: '~> 3.0.0',
+          newValue: '~> 3.37.0',
```

Comment:
```suggestion
          newValue: '~> 3.37',
```
Just to test the number full version is templated.